Add trip detail endpoint

aleksandr-hilko opened the pull request:
In this PR:
1. created endpoints:
 - GET, PUT, DELETE api/trips/{trip_id}
 - POST api/trips/{trip_id}/reserve/
 - GET api/trips/{trip_id}/requests/
For more details please look at readme file that I updated with the description of api endpoints (the newly added correspond to the 3-7 points in the 'Description of API endpoints' section).
2. Added test that cover newly added endpoints.
3. Added TripRequest serializer class.

aleksandr-hilko committed:
add trip detail endpoint and tests

aleksandr-hilko committed:
add api/trip/<trip_id>/reserve/ and /api/trips/<trip_id>/requests/ endpoints.

aleksandr-hilko committed:
add tests for /api/trips/<id>/reserve/ endpoint

aleksandr-hilko committed:
add permission tests

aleksandr-hilko committed:
add 'free_seats' field to serializer

aleksandr-hilko committed:
add endpoints information to the readme file

aleksandr-hilko commented on trip_app/conftest.py:
```diff
@@ -0,0 +1,27 @@
+def create_user(signup_data, client):
+    """" Create a new user and return user id, username and user password """
+    password = signup_data.pop('password')
+    user = User.objects.create_user(username=signup_data["username"], password=password)
```
This was just replaced from accounts/tests/conftest folder to the root conftest of a project to be shared across the tests in different apps.

Meathookseed commented on README.md:
```diff
@@ -28,4 +28,107 @@ App that allow customers to collaborate being moved around the city by either cr
     You can check that your server is now working in a browser:
         `http://127.0.0.1:8000/api/accounts/`
     Press `Ctrl+C` when completed to stop the services.
+    
+ ## Description of API endpoints
```
Its better to put all this comments under docstring directly in views. But you should filter out some information out of here ( F.E. information about start_point and dest_point.) This kind of information should be placed directly under func or class which operates with information..

aleksandr-hilko replied:
Actually I added this API documentation here in order to have some sort of requirements in one place without having people to dig into code of the functions and classes to get some initial understanding of what's going on. I realize that for people who are dealing with geo coordinates for the first time it might be difficult to realize what sort of input should they send in request... But yeah I understand that my functions lack of doc strings, will try to fullfil them with documentation more ;)

Meathookseed replied:
You should understand that almost all people who would read your code - developers. And there is no need to explain some obvious things that man can understand from signature of a function. Especially if function/class has good/brief documentation about what it does.

Meathookseed commented on trip_app/accounts/permissions.py:
```diff
@@ -3,10 +3,8 @@
 class IsAuthOwnerOrAdminOnly(permissions.IsAuthenticated):
     """
-    Custom permission to only allow owners and admins to get/edit it.
+    Permissions to allow only owners and admins to get/edit/delete a user
```
Comment on first line of docstring and dot!
```
def my_func():
    """Do the thing."""
```

Meathookseed commented on trip_app/conftest.py:
```diff
@@ -0,0 +1,27 @@
+
+@pytest.fixture
+def create_user(signup_data, client):
+    """" Create a new user and return user id, username and user password """
```
First line of docstring says only about what function does. If you want provide it with information about what returns you can do like this:
```
def my_func(my_param):
    """Do the thing.

    :return: String representing what returns.
    """
```

aleksandr-hilko replied:
Should be updated now. Thanks.

Meathookseed commented on trip_app/conftest.py:
```diff
@@ -0,0 +1,27 @@
+def create_user(signup_data, client):
+    """" Create a new user and return user id, username and user password """
+    password = signup_data.pop('password')
+    user = User.objects.create_user(username=signup_data["username"], password=password)
```
I think you can create user just like this `User.objects.create_user(**signup_data)`

aleksandr-hilko replied:
Yeah this might work, Thanks!

Meathookseed commented on trip_app/trips/models.py:
```diff
@@ -16,8 +16,12 @@ class Trip(models.Model):
     created = models.DateTimeField(auto_now_add=True)
 
+    @property
+    def free_seats(self):
```
It's better that every function have docstring, even if it's easy to understand what does function implements.

aleksandr-hilko replied:
Added doc string. Thanks.

Meathookseed commented on trip_app/trips/models.py:
```diff
@@ -16,8 +16,12 @@ class Trip(models.Model):
-    trip = models.ForeignKey(Trip, on_delete=models.CASCADE)
-    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    trip = models.ForeignKey(Trip, on_delete=models.CASCADE, related_name="requests")
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="requests")
```
Limit all lines to a maximum of 79 characters.

andrey-bulygin replied:
That's a matter of project convention usually, but I also prefer this `79` limit option.

aleksandr-hilko replied:
I realize that PEP8 suggests using 79, but for me it makes files too long, though it might be the matter of habbit. I'll set the pre-commit black hook to format the files limiting 79 lines.

Meathookseed commented on trip_app/trips/permissions.py:
```diff
@@ -0,0 +1,10 @@
+
+class IsTripDriverOrAdmin(IsAuthenticated):
+    """
+    Permissions to allow only driver of a trip or admin to edit/delete a trip
```
Sentence should be on a first line of docstring like this : `"""Permissions to allow only driver of a trip or admin to edit/delete a trip."""`  and always dot at the end.

andrey-bulygin replied:
This is too tough in my opinion. Typically we'd start description from the next line at most of projects. @Meathookseed any specific ideas why the first line is better?

Meathookseed replied:
https://www.python.org/dev/peps/pep-0257/

Meathookseed commented on trip_app/trips/tests/test_permissions.py:
```diff
@@ -0,0 +1,63 @@
+from django.contrib.auth import get_user_model
+from django.urls import reverse
+
+from .trip_factory import TripFactory
```
Its better to determine what style of imports you use, relative or direct. 
Across all project.

andrey-bulygin replied:
I'd advocate absolute imports everywhere.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -1,23 +1,33 @@
+            permission_classes = [IsTripDriverOrAdmin]
+        else:
+            permission_classes = [IsAuthenticated]
+        return [permission() for permission in permission_classes]
```
This could be changed to: 
```
def get_permissions(self):
    if self.action in ['update', 'destroy', 'partial_update']:
            self.permission_classes = [IsTripDriverOrAdmin]
        else:
            self.permission_classes = [IsAuthenticated]
        return super().get_permissions()

```

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -1,23 +1,33 @@
+
     def get_queryset(self):
         queryset = Trip.objects.all()
         query_params = self.request.query_params
```
Rename `sp` and `dp`  variables to human readable format.

andrey-bulygin replied:
Not sure where it relates to (some github formatting issue for me, I believe), but I agree: maybe `point1 + point2` the way you handle times?

aleksandr-hilko replied:
Renamed to start_point and destination_point. Thanks.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -43,3 +53,30 @@ def get_queryset(self):
+
+    @action(detail=True, methods=['post'])
+    def reserve(self, request, pk=None):
+        """ Reserve a trip by sending a POST to api/trip/<trip_id>/reserve/.
```
Format docstring. Delete unnecessary information from it.

aleksandr-hilko replied:
Genuine question, what information do you consider unnecessary in this doc string ?

Meathookseed replied:
There is no need in this : ``` If the man_approve field of the trip model is True the trip request will be created,
which means that user requested a trip should be approved by the driver.
If man_approve is False user will be added to the passengers list if there are free seats ``` because all this information can be understood from code below.

aleksandr-hilko replied:
Removed mentioned comments.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -43,3 +53,30 @@ def get_queryset(self):
+        trip = self.get_object()
+        user = request.user
+        if user == trip.driver:
+            return Response(data="Driver can't be a passenger at the same time", status=status.HTTP_400_BAD_REQUEST)
```
Limit all lines to a maximum of 79 characters.

aleksandr-hilko replied:
Should be limited now. Thanks.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -43,3 +53,30 @@ def get_queryset(self):
+            if trip.man_approve:
+                trip_request = TripRequest.objects.create(trip=trip, user=user)
+                serializer = TripRequestSerializer(trip_request)
+                return Response(serializer.data)
```
Why we need data on response, maybe just 201 code?

aleksandr-hilko replied:
Hm. I'm not sure what is the best practice for such cases, I used to think that it is recommended to return the entity in the response body when smth is created on the server, though indeed the response 201 should be more appropriate then 200

aleksandr-hilko replied:
According to the specification of 201 response

> The response SHOULD include an entity containing a list of resource characteristics and location(s) from which the user or user agent can choose the one most appropriate. 

So I guess should either include the representation of resource created or link to the resource in headers. What do you think about it ?

andrey-bulygin replied:
My vision is it's better to return the entity here. 201 is probably a bit better. E.g. you could have the same endpoint create/update a resource (just believe me, lol), and 200 vs 201 would make a great difference there.

andrey-bulygin replied:
any final choice here? @aleksandr-hilko

aleksandr-hilko replied:
I switched here to 201 + entity in response.

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -43,3 +53,30 @@ def get_queryset(self):
+                return Response(serializer.data)
+            trip.passengers.add(user)
+            serializer = self.get_serializer(trip)
+            return Response(serializer.data)
```
Same here.

aleksandr-hilko replied:
I'm leaning towards 200 response here because in this case no resource is created but passenger is just added to the trip passengers list, and I return the trip instance in this case.

andrey-bulygin replied:
the main item here is the resource (not the underlying models and/or relations), so 201 makes sense here. That said, it's up to your general choice of a status for all created resources throughout your API. Keep it 200 everywhere if you prefer so.

aleksandr-hilko replied:
Thanks for the clarification @andreybulygin. I'm okay with returning either 201/200 responses. But I'm afraid that still got some confusion on this. Let me explain my view on this and try to advocate my point:
So within this particular endpoint I want to reserve a trip. And 2 cases might happen: 
1) if a trip can be auto reserved I just add the user sent the request to the passengers list. And in this case I think that 200 is more relevant, because resource (which is trip in this case) is not created as 201 really assumes, it is updated. And to show that trip resource is updated I add the trip resource in the response. 
2) If a trip assumes an approval from driver, trip request resource is created. And in this case I want to return 201 + trip request entity in the response.
My intuition says me that this is the correct way to do this. Though may be it is correct to stay 201 with POST and 200 with PUT to prevent messing things.

Meathookseed replied:
REST doesn't have any strict and MUST rules in spec, and thats is veeery big minus of it. So be comfort to use what you think is correct.

Meathookseed reviewed (commented):
Comments below

andrey-bulygin commented on trip_app/trips/tests/test_views.py:
```diff
@@ -42,66 +46,159 @@ def trips_with_different_coordinates():
-    assert resp_dict["count"] == 1
-    assert resp_dict["results"][0]["id"] == in_time_trip.id
+class TestTrips:
+    def test_create_trip(self, admin_client, trip_data):
```
I'd say that `trip` is more or less redundant in the method names (as `TestTrips` already suggests that)

aleksandr-hilko replied:
I would agree, in this case, though there might be cases where being more explicit might be justified,
e.g. for GET it is common to have endpoints for getting list and individual entry. In this case I think including trip in function name `test_get_trip` should be appreciated even is there is no GET /trips/ endpoint. And it brings up the question: should I be consistent across the test names in the same file which means to include 'trip' part of names everywhere or to include it only where it might be perceived ambiguous ? I haven't got strong opinion on this.....

andrey-bulygin replied:
I'd follow the approach:
1. if namespace (class name) defines clearly what we are working on, shorten the method names 
1. do specify extra name suffix in more specific cases
Anyway, it makes sense to keep reasonably short method names wherever possible.

aleksandr-hilko replied:
Thanks @andreybulygin. I changed names of the tests in this file according to this approach.

aleksandr-hilko committed:
Addressed PR comments

Fixed doc strings, limited line length up to 79 characters.
Unified imports to using only absolute. Added pre commit black hook.

aleksandr-hilko committed:
small format fix

Meathookseed commented on trip_app/conftest.py:
```diff
@@ -0,0 +1,31 @@
+        """
+    password = signup_data.get("password")
+    user = User.objects.create_user(**signup_data)
+    return user.id, user.username, password
```
Why you return password not from `user.password` - according to your code, there is no any hashing of password when setting password to user, so why not to do this?

aleksandr-hilko replied:
Actually `_create_user` which is being called by `create_user` of `django/contrib/auth/models.UserManager` class hashes the password and `user.password` will return hashed value, and I need raw value to get the token.

Meathookseed commented on trip_app/trips/tests/conftest.py:
```diff
@@ -17,10 +19,12 @@ def trips(request):
-    del data["passengers"]
+    data = {
+        "dep_time": (timezone.now() + timedelta(days=1)),
+        "start_point": f"{faker.latitude()} {float(faker.longitude())}",
```
Could you explain why longitude cast to float and latitude not?
I checked, you can use `faker.latlng` function instead generating this individually.

aleksandr-hilko replied:
It is just copypast from me, sorry. Actually there is no need to cast to float in this case as I need str here and f-string would call the str method of each decimal with proper formatting, therefore I will remove cast at all.
Regarding `faker.latlng` it returns tuple of Decimal objects, which, I'm now sure how to convert to proper string that I need (though if you know how to do it please guide me with more details).

Meathookseed commented on trip_app/trips/models.py:
```diff
@@ -4,8 +4,14 @@
+    )
     dep_time = models.DateTimeField()
     start_point = geo_models.PointField()
     dest_point = geo_models.PointField()
```
Its better to use `decimal` type for any price/currency/money values. This type is more precise on calculations with floating points, so i suggest you to change this.

aleksandr-hilko replied:
Good point! Changed this field to decimal type. Thanks.

Meathookseed commented on trip_app/trips/tests/conftest.py:
```diff
@@ -17,10 +19,12 @@ def trips(request):
+        "dep_time": (timezone.now() + timedelta(days=1)),
+        "start_point": f"{faker.latitude()} {float(faker.longitude())}",
+        "dest_point": f"{faker.latitude()} {float(faker.longitude())}",
+        "price": faker.random_number(),
```
If you change price to `decimal` type, do not forget to change code here too.

aleksandr-hilko replied:
Changed here as well. Thanks.

Meathookseed commented on trip_app/trips/tests/trip_factory.py:
```diff
@@ -1,35 +1,41 @@
-    start_point = Point(float(faker.latitude()), float(faker.longitude()), srid=4326)
-    dest_point = Point(float(faker.latitude()), float(faker.longitude()), srid=4326)
+    start_point = Point(
+        float(faker.latitude()), float(faker.longitude()), srid=4326
```
And here you cast both - lat and long to float

aleksandr-hilko replied:
Here I have to cast in both cases as  `faker.latitude()` returns Decimal object but the initializer of Point class expect it to be only `int` or `float` and if I didn't do the cast, the TypeError will be raised.

Meathookseed commented on trip_app/trips/tests/test_views.py:
```diff
@@ -1,19 +1,23 @@
 from datetime import timedelta
```
Look at https://testtools.readthedocs.io/en/latest/ , this can be useful for tests, have a lot of useful types of asserts and a lot yummy thing in there.

aleksandr-hilko replied:
Thanks, I will take a look at this!

Meathookseed commented on trip_app/trip_app/settings/base.py:
```diff
@@ -31,131 +31,127 @@ def get_env_variable(var_name):
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = 'gdgh$)yv7v8*wuiw^0hk5l(9e57&=b1giij#5hq6=$5732y=95'
+SECRET_KEY = "gdgh$)yv7v8*wuiw^0hk5l(9e57&=b1giij#5hq6=$5732y=95"
```
Hide this to .env/.conf/keys.yaml or to something like jks. Choose what you like.

aleksandr-hilko replied:
Thanks. Removed this to the env variables.

aleksandr-hilko committed:
change to decimal type  price field of trip model

aleksandr-hilko committed:
removed secret key from settings

Meathookseed commented on trip_app/trips/views.py:
```diff
@@ -1,45 +1,91 @@
         time2 = query_params.get("time2")
-        sp = query_params.get("sp")
-        dp = query_params.get("dp")
+        start_point = query_params.get("sp")
```
I don't get why you sending coordinates like this `coord1 coord2`, but not `coord1, coord2`. It is easier to parse values then from regular string, and then in `str_to_geopoint` you do not need any regex to parse something. Just checked with gmaps, and they use comma between `x,y` coordinate. Explain me why you chose such style if there is any reason for this 🙃

aleksandr-hilko replied:
Well... initially I wasn't sure what are the rules and formats therefore decided to use re here to expand the range of incoming formats of data. 
I expected the clients would send us coordinates in string formats such as '23.4 23.5', '[23.4 23.5]', '(23.4 23.5)'.. 
Not it looks weird though, but I'm planing to address it in next PR and add drf-gis library that contains proper serializers to be able to get rid of `str_to_geopoint` and in general clear things up.

Meathookseed reviewed (approved):
I'm happy enough for now with changes, thanks for patience :)

aleksandr-hilko commented:
> I'm happy enough for now with changes, thanks for patience :)

And thanks for your patience :)
